### ISSUE TITLE
GCE PD Data Loss

### ISSUE THREAD

**Issue by smreed** [GoogleCloudPlatform/kubernetes#3965]
0.9.1 (edit: originally thought it was 0.9.2) in GCE and I'm seeing data disappear from GCE PDs after deleting the pod that mounts it.

I'm attaching some relevant looking bits from the kubelet log.

```
I0130 01:00:21.156131    4820 kubelet.go:1255] Killing unwanted container {podFullName:prometheus.default.api uid:6fcd7fc4-a7ff-11e4-85c6-42010af0b3db containerName:prometheus}
I0130 01:00:21.156169    4820 kubelet.go:863] Killing container with id "1f85b85dc825b218f9273869678d09b7f6c3a93307ac2b5edb9255f5a406dfb3" and name "/k8s_prometheus.95c0e9de_prometheus.default.api_6fcd7fc4-a7ff-11e4-85c6-42010af0b3db_0f4a949a"
I0130 01:00:21.156331    4820 event.go:117] Event(api.ObjectReference{Kind:"BoundPod", Namespace:"default", Name:"prometheus", UID:"6fcd7fc4-a7ff-11e4-85c6-42010af0b3db", APIVersion:"v1beta1", ResourceVersion:"", FieldPath:"spec.containers{node-exporter}"}): reason: 'killing' Killing b4bfcd6c3de50df86f1e6dfb5ddbcba94fb731c07375f849e907d685bde1c142 - /k8s_node-exporter.562dc0b8_prometheus.default.api_6fcd7fc4-a7ff-11e4-85c6-42010af0b3db_6354e686
I0130 01:00:21.322846    4820 event.go:117] Event(api.ObjectReference{Kind:"BoundPod", Namespace:"default", Name:"prometheus", UID:"6fcd7fc4-a7ff-11e4-85c6-42010af0b3db", APIVersion:"v1beta1", ResourceVersion:"", FieldPath:"spec.containers{prometheus}"}): reason: 'killing' Killing 1f85b85dc825b218f9273869678d09b7f6c3a93307ac2b5edb9255f5a406dfb3 - /k8s_prometheus.95c0e9de_prometheus.default.api_6fcd7fc4-a7ff-11e4-85c6-42010af0b3db_0f4a949a
E0130 01:00:21.354926    4820 kubelet.go:1345] Couldn't sync containers: remove /var/lib/kubelet/pods/6fcd7fc4-a7ff-11e4-85c6-42010af0b3db/volumes/kubernetes.io~gce-pd/prometheus-storage: device or resource busy
E0130 01:00:31.359134    4820 kubelet.go:1345] Couldn't sync containers: remove /var/lib/kubelet/pods/6fcd7fc4-a7ff-11e4-85c6-42010af0b3db/volumes/kubernetes.io~gce-pd/prometheus-storage: device or resource busy
E0130 01:00:41.363139    4820 kubelet.go:1345] Couldn't sync containers: remove /var/lib/kubelet/pods/6fcd7fc4-a7ff-11e4-85c6-42010af0b3db/volumes/kubernetes.io~gce-pd/prometheus-storage: device or resource busy
E0130 01:00:51.368149    4820 kubelet.go:1345] Couldn't sync containers: remove /var/lib/kubelet/pods/6fcd7fc4-a7ff-11e4-85c6-42010af0b3db/volumes/kubernetes.io~gce-pd/prometheus-storage: device or resource busy
E0130 01:01:01.372352    4820 kubelet.go:1345] Couldn't sync containers: remove /var/lib/kubelet/pods/6fcd7fc4-a7ff-11e4-85c6-42010af0b3db/volumes/kubernetes.io~gce-pd/prometheus-storage: device or resource busy
```

Around this point I ssh to the minion and `ls -l` the root of the PD and found nothing there. There was definitely data on the device before I deleted the pod.

Also, after this happens, I have to manually detach the disk using `gcloud` or the GCE console and then `umount` all of the places where the PD was mounted on the minion.

**Comment by smreed** [comment]
This pod is deployed stand-alone, not by a replication controller, in case that makes a difference. I've seen it with other pods as well, such as one running mysql.

**Comment by brendandburns** [comment]
Thanks for the report, I will try to look into it.

Can you grab the event stream (kubectl get events) next time it happens as
well?

Thanks
--brendan

**Comment by thockin** [comment]
Do you think this was related to the PR you just put in with RemoveAll?
This is why I suggested maybe re-checking isMountPoint() and ONLY
proceeding with Remove() if it is not...

**Comment by smreed** [comment]
I'll grab events next time. I can repro this easily so I'll try to do it
first thing tomorrow.

**Comment by zmerlynn** [comment]
We really need an actual persistent data e2e :/

**Comment by smreed** [comment]
These are the only events I see from `kubectl get events` from when I killed the pod. 

```
Fri, 30 Jan 2015 01:00:21 +0000   prometheus                                           BoundPod            spec.containers{node-exporter}              killing             {kubelet kubernetes-minion-7.c.eternal-empire-754.internal}    Killing b4bfcd6c3de50df86f1e6dfb5ddbcba94fb731c07375f849e907d685bde1c142 - /k8s_node-exporter.562dc0b8_prometheus.default.api_6fcd7fc4-a7ff-11e4-85c6-42010af0b3db_6354e686
Fri, 30 Jan 2015 01:00:21 +0000   prometheus                                           BoundPod            implicitly required container net           killing             {kubelet kubernetes-minion-7.c.eternal-empire-754.internal}    Killing 7b850cb0c0ddf83a8a9772ae8f58a981ce77ffe1ed536ba08bd20882b8485cd7 - /k8s_net.ba9fd962_prometheus.default.api_6fcd7fc4-a7ff-11e4-85c6-42010af0b3db_925c3fed
Fri, 30 Jan 2015 01:00:21 +0000   prometheus                                           BoundPod            spec.containers{prometheus}                 killing             {kubelet kubernetes-minion-7.c.eternal-empire-754.internal}    Killing 1f85b85dc825b218f9273869678d09b7f6c3a93307ac2b5edb9255f5a406dfb3 - /k8s_prometheus.95c0e9de_prometheus.default.api_6fcd7fc4-a7ff-11e4-85c6-42010af0b3db_0f4a949a
```

**Comment by smreed** [comment]
One more repro, with hopefully more clues.

Pod is deployed, PD is attached properly, and data is being persisted.

    root@kubernetes-minion-2:~# find /var/lib/kubelet/plugins/kubernetes.io/gce-pd/mounts/prometheus-storage | wc -l
    890
    root@kubernetes-minion-2:~# du -h --max-depth 0 /var/lib/kubelet/plugins/kubernetes.io/gce-pd/mounts/prometheus-storage
    26M     /var/lib/kubelet/plugins/kubernetes.io/gce-pd/mounts/prometheus-storage

Now from my laptop I `kubectl delete pod prometheus`. Here's the output from `kubectl get events` shortly thereafter.

```
Fri, 30 Jan 2015 16:13:39 +0000   prometheus                                           BoundPod            implicitly required container net           killing             {kubelet kubernetes-minion-2.c.eternal-empire-754.internal}    Killing abdbe3ad74c4ac77e2d91f97a65afb0e84e1175b4c72deaa65ff11e8687523b6 - /k8s_net.ba9fd962_prometheus.default.api_91105043-a81b-11e4-85c6-42010af0b3db_385031d6
Fri, 30 Jan 2015 16:13:39 +0000   prometheus                                           BoundPod            spec.containers{node-exporter}              killing             {kubelet kubernetes-minion-2.c.eternal-empire-754.internal}    Killing 7afd3d877c53b0daa17b1339594a718f242fe798ef6eff74d92661d02a4e7a20 - /k8s_node-exporter.562dc0b8_prometheus.default.api_91105043-a81b-11e4-85c6-42010af0b3db_d04b2558
Fri, 30 Jan 2015 16:13:39 +0000   prometheus                                           BoundPod            spec.containers{prometheus}                 killing             {kubelet kubernetes-minion-2.c.eternal-empire-754.internal}    Killing ec41aaaab1f3af969735728699f91e026f1d69ff50edafb4663d3fe58ad03504 - /k8s_prometheus.95c0e9de_prometheus.default.api_91105043-a81b-11e4-85c6-42010af0b3db_10566093
```

If I run `gcloud compute instances describe kubernetes-minion-2` here's the "disks" section:

```
disks:
- autoDelete: true
  boot: true
  deviceName: persistent-disk-0
  index: 0
  interface: SCSI
  kind: compute#attachedDisk
  mode: READ_WRITE
  source: https://www.googleapis.com/compute/v1/projects/eternal-empire-754/zones/us-central1-f/disks/kubernetes-minion-2
  type: PERSISTENT
- deviceName: prometheus-storage
  index: 1
  interface: SCSI
  kind: compute#attachedDisk
  mode: READ_WRITE
  source: https://www.googleapis.com/compute/v1/projects/eternal-empire-754/zones/us-central1-f/disks/prometheus-storage
  type: PERSISTENT
```

The disk is still attached, what about the contents of the disk?

```
root@kubernetes-minion-2:~# df -h | grep prom
/dev/disk/by-id/google-prometheus-storage               197G   60M  187G   1% /var/lib/kubelet/plugins/kubernetes.io/gce-pd/mounts/prometheus-storage
/dev/disk/by-id/google-prometheus-storage               197G   60M  187G   1% /var/lib/kubelet/pods/91105043-a81b-11e4-85c6-42010af0b3db/volumes/kubernetes.io~gce-pd/prometheus-storage

root@kubernetes-minion-2:~# du -h --max-depth 0 /var/lib/kubelet/plugins/kubernetes.io/gce-pd/mounts/prometheus-storage
4.0K    /var/lib/kubelet/plugins/kubernetes.io/gce-pd/mounts/prometheus-storage

root@kubernetes-minion-2:~# find /var/lib/kubelet/plugins/kubernetes.io/gce-pd/mounts/prometheus-storage | wc -l                                                             
1
```

Finally the output of `kubectl version`. Looks like I lied about running 0.9.2, we're running 0.9.1 on the servers (I'll update my original comment).

```
Running: ./../cluster/gce/../../platforms/linux/amd64/kubectl version
Client Version: version.Info{Major:"0", Minor:"9+", GitVersion:"v0.9.2-dirty", GitCommit:"013a97ea6e2a143f763035defdbcbcdc1b6d3411", GitTreeState:"dirty"}
Server Version: &version.Info{Major:"0", Minor:"9+", GitVersion:"v0.9.1-dirty", GitCommit:"3623a01bf0e90de6345147eef62894057fe04b29", GitTreeState:"dirty"}
```

**Comment by smreed** [comment]
@brendandburns it seems like your commit https://github.com/GoogleCloudPlatform/kubernetes/commit/a0b8429d2e57e15d559333a79f33ff58ad1cf34e would fix this. Looks like the old code was pretty clearly blowing everything away with `os.RemoveAll` before asking for the PD to be detached.

**Comment by thockin** [comment]
It SHOULD have been unmounted first, so there is still an open question of
why the data was visible there.  Maybe we should also change that RemoveAll
to a simple rmdir, which will fail if it is not empty.  Not sure if I can
rmdir a mount point, but I think not..

**Comment by brendandburns** [comment]
Yeah, I'll update https://github.com/GoogleCloudPlatform/kubernetes/pull/3967 to just try to rmdir...

**Comment by brendandburns** [comment]
Ok, #3967 is merged. @smreed can you try to repro?

Thanks!

**Comment by smreed** [comment]
Probably not until tomorrow, but I will ASAP for sure.

**Comment by smreed** [comment]
Deployed 0.10.0 and tested this.

kube-down did not blow away any data on the disk, it was present after the new cluster was up and the pods were deployed.

Then, I deleted the pod w/ the PD. I saw this repeated in the kubelet.log:

    E0204 21:34:15.454164    4722 kubelet.go:1359] Couldn't sync containers: remove /var/lib/kubelet/pods/3faea999-acb3-11e4-9b3d-42010af0f1f5/volumes/kubernetes.io~gce-pd/prometheus-storage: device or resource busy

The GCE console reported the disk as still being attached to the minion. The pod containers were not running.

Ran `ls -al /var/lib/kubelet/plugins/kubernetes.io/gce-pd/mounts/prometheus-storage`, the directory was empty.

Deployed the pod again, it landed on a different minion but stayed pending due to the fact that the disk was already attached elsewhere.

I manually detached the PD from the old minion, and the pod eventually deployed on the new one. I confirmed on this minion that the PD came in empty with no data.

kubectl version:
    Client Version: version.Info{Major:"0", Minor:"10+", GitVersion:"v0.10.0-dirty", GitCommit:"71e26cbeb9c0bf1b6cc4c29b1f9d930e53513911", GitTreeState:"dirty"}
    Server Version: version.Info{Major:"0", Minor:"10+", GitVersion:"v0.10.0-dirty", GitCommit:"71e26cbeb9c0bf1b6cc4c29b1f9d930e53513911", GitTreeState:"dirty"}

Adding output of `kubectl events` for this pod in the next comment.

**Comment by smreed** [comment]
```
Wed, 04 Feb 2015 21:17:39 +0000   prom-dash                                            BoundPod            implicitly required container POD           pulled              {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Successfully pulled image "kubernetes/pause:latest"
Wed, 04 Feb 2015 21:17:39 +0000   prometheus-push                                      BoundPod            implicitly required container POD           pulled              {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Successfully pulled image "kubernetes/pause:latest"
Wed, 04 Feb 2015 21:17:39 +0000   prom-dash                                            Pod                                                             scheduled           {scheduler }                                                   Successfully assigned prom-dash to kubernetes-minion-3.c.eternal-empire-754.internal
Wed, 04 Feb 2015 21:17:39 +0000   prom-dash                                            BoundPod            implicitly required container POD           created             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Created with docker id efcf5e1e2ee0e8988de356ec5adedeb28b6969b23ed60f0339e7b36a554fe8dd
Wed, 04 Feb 2015 21:17:39 +0000   prom-dash                                            BoundPod            implicitly required container POD           started             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Started with docker id efcf5e1e2ee0e8988de356ec5adedeb28b6969b23ed60f0339e7b36a554fe8dd
Wed, 04 Feb 2015 21:17:39 +0000   prometheus-push                                      Pod                                                             scheduled           {scheduler }                                                   Successfully assigned prometheus-push to kubernetes-minion-3.c.eternal-empire-754.internal
Wed, 04 Feb 2015 21:17:39 +0000   prometheus-push                                      BoundPod            implicitly required container POD           created             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Created with docker id 2e07f817cf2aabf7e55ee57447e65a4d8bc2ca164dce9ff6353b18163af76803
Wed, 04 Feb 2015 21:17:40 +0000   prometheus-push                                      BoundPod            implicitly required container POD           started             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Started with docker id 2e07f817cf2aabf7e55ee57447e65a4d8bc2ca164dce9ff6353b18163af76803
Wed, 04 Feb 2015 21:17:40 +0000   prometheus                                           Pod                                                             scheduled           {scheduler }                                                   Successfully assigned prometheus to kubernetes-minion-3.c.eternal-empire-754.internal
Wed, 04 Feb 2015 21:17:40 +0000   prometheus                                           BoundPod            implicitly required container POD           started             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Started with docker id f38347692abc3e75ff5652bb608bdde04cd08a990c2587ff5e917b2c0633f84f
Wed, 04 Feb 2015 21:17:40 +0000   prometheus                                           BoundPod            implicitly required container POD           created             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Created with docker id f38347692abc3e75ff5652bb608bdde04cd08a990c2587ff5e917b2c0633f84f
Wed, 04 Feb 2015 21:17:40 +0000   prometheus                                           BoundPod            implicitly required container POD           pulled              {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Successfully pulled image "kubernetes/pause:latest"
Wed, 04 Feb 2015 21:18:46 +0000   prometheus-push                                      BoundPod            spec.containers{prom-pushgateway}           started             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Started with docker id 83a4e9d913731cce6c6b6add9877aaa389c409965a80d026d50c212abcac9f8a
Wed, 04 Feb 2015 21:18:46 +0000   prometheus-push                                      BoundPod            spec.containers{prom-pushgateway}           created             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Created with docker id 83a4e9d913731cce6c6b6add9877aaa389c409965a80d026d50c212abcac9f8a
Wed, 04 Feb 2015 21:18:46 +0000   prometheus-push                                      BoundPod            spec.containers{prom-pushgateway}           pulled              {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Successfully pulled image "fish/prometheus-pushgateway"
Wed, 04 Feb 2015 21:19:55 +0000   prom-dash                                            BoundPod            spec.containers{prom-dash-db}               created             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Created with docker id a9f164c8264b2dd63f5788bc9a192ab32c519e7af41eea5614e82e5b9c1590ab
Wed, 04 Feb 2015 21:19:55 +0000   prom-dash                                            BoundPod            spec.containers{prom-dash-db}               pulled              {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Successfully pulled image "gcr.io/eternal_empire_754/core-prom-dash-db:latest"
Wed, 04 Feb 2015 21:19:55 +0000   prom-dash                                            BoundPod            spec.containers{prom-dash-db}               started             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Started with docker id a9f164c8264b2dd63f5788bc9a192ab32c519e7af41eea5614e82e5b9c1590ab
Wed, 04 Feb 2015 21:20:04 +0000   prometheus                                           BoundPod            spec.containers{prometheus}                 started             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Started with docker id 981ae37c7e2d71993e0439e4b955e0ef826684520938a6ad2cad52b03b83a6e0
Wed, 04 Feb 2015 21:20:04 +0000   prometheus                                           BoundPod            spec.containers{prometheus}                 pulled              {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Successfully pulled image "gcr.io/eternal_empire_754/core-prometheus-server:latest"
Wed, 04 Feb 2015 21:20:04 +0000   prometheus                                           BoundPod            spec.containers{prometheus}                 created             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Created with docker id 981ae37c7e2d71993e0439e4b955e0ef826684520938a6ad2cad52b03b83a6e0
Wed, 04 Feb 2015 21:21:11 +0000   prometheus                                           BoundPod            spec.containers{node-exporter}              started             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Started with docker id 84f3ebb3d0d650180deeb8568e6f9b9ddad07955c300917eb964d07b1faf8da9
Wed, 04 Feb 2015 21:21:11 +0000   prometheus                                           BoundPod            spec.containers{node-exporter}              pulled              {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Successfully pulled image "prom/node-exporter:latest"
Wed, 04 Feb 2015 21:21:11 +0000   prometheus                                           BoundPod            spec.containers{node-exporter}              created             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Created with docker id 84f3ebb3d0d650180deeb8568e6f9b9ddad07955c300917eb964d07b1faf8da9
Wed, 04 Feb 2015 21:21:41 +0000   prom-dash                                            BoundPod            spec.containers{prom-dash}                  started             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Started with docker id 9ed55637fa95cfef21daf59ae2fb221fa9c7e366160913428996d27714a1de10
Wed, 04 Feb 2015 21:21:41 +0000   prom-dash                                            BoundPod            spec.containers{prom-dash}                  created             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Created with docker id 9ed55637fa95cfef21daf59ae2fb221fa9c7e366160913428996d27714a1de10
Wed, 04 Feb 2015 21:21:41 +0000   prom-dash                                            BoundPod            spec.containers{prom-dash}                  pulled              {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Successfully pulled image "prom/promdash:latest"
Wed, 04 Feb 2015 21:33:40 +0000   prometheus                                           BoundPod            implicitly required container POD           killing             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Killing f38347692abc3e75ff5652bb608bdde04cd08a990c2587ff5e917b2c0633f84f - /k8s_POD.ebe10168_prometheus.default.api_3faea999-acb3-11e4-9b3d-42010af0f1f5_9e4499f8
Wed, 04 Feb 2015 21:33:40 +0000   prometheus                                           BoundPod            spec.containers{node-exporter}              killing             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Killing 84f3ebb3d0d650180deeb8568e6f9b9ddad07955c300917eb964d07b1faf8da9 - /k8s_node-exporter.708ae69c_prometheus.default.api_3faea999-acb3-11e4-9b3d-42010af0f1f5_d5b47239
Wed, 04 Feb 2015 21:33:41 +0000   prometheus                                           BoundPod            spec.containers{prometheus}                 killing             {kubelet kubernetes-minion-3.c.eternal-empire-754.internal}    Killing 981ae37c7e2d71993e0439e4b955e0ef826684520938a6ad2cad52b03b83a6e0 - /k8s_prometheus.88d40fd1_prometheus.default.api_3faea999-acb3-11e4-9b3d-42010af0f1f5_86124870
Wed, 04 Feb 2015 21:34:03 +0000   prometheus                                           Pod                                                             scheduled           {scheduler }                                                   Successfully assigned prometheus to kubernetes-minion-8.c.eternal-empire-754.internal
Wed, 04 Feb 2015 21:34:03 +0000   prometheus                                           BoundPod            implicitly required container POD           created             {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Created with docker id 3d3421caa75643055cd2335e4018c3baf1e0f924b3e00674c2e16434f98a9c58
Wed, 04 Feb 2015 21:34:03 +0000   prometheus                                           BoundPod            implicitly required container POD           pulled              {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Successfully pulled image "kubernetes/pause:latest"
Wed, 04 Feb 2015 21:34:03 +0000   prometheus                                           BoundPod            implicitly required container POD           started             {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Started with docker id 3d3421caa75643055cd2335e4018c3baf1e0f924b3e00674c2e16434f98a9c58
Wed, 04 Feb 2015 21:34:13 +0000   prometheus                                           BoundPod                                                        failedSync          {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Error syncing pod, skipping: Could not attach disk: Timeout after 10s
Wed, 04 Feb 2015 21:34:33 +0000   prometheus                                           BoundPod                                                        failedSync          {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Error syncing pod, skipping: Could not attach disk: Timeout after 10s
Wed, 04 Feb 2015 21:34:43 +0000   prometheus                                           BoundPod                                                        failedSync          {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Error syncing pod, skipping: Could not attach disk: Timeout after 10s
Wed, 04 Feb 2015 21:35:03 +0000   prometheus                                           BoundPod                                                        failedSync          {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Error syncing pod, skipping: Could not attach disk: Timeout after 10s
Wed, 04 Feb 2015 21:35:13 +0000   prometheus                                           BoundPod                                                        failedSync          {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Error syncing pod, skipping: Could not attach disk: Timeout after 10s
Wed, 04 Feb 2015 21:35:23 +0000   prometheus                                           BoundPod                                                        failedSync          {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Error syncing pod, skipping: Could not attach disk: Timeout after 10s
Wed, 04 Feb 2015 21:35:43 +0000   prometheus                                           BoundPod                                                        failedSync          {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Error syncing pod, skipping: Could not attach disk: Timeout after 10s
Wed, 04 Feb 2015 21:35:53 +0000   prometheus                                           BoundPod                                                        failedSync          {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Error syncing pod, skipping: Could not attach disk: Timeout after 10s
Wed, 04 Feb 2015 21:36:03 +0000   prometheus                                           BoundPod                                                        failedSync          {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Error syncing pod, skipping: Could not attach disk: Timeout after 10s
Wed, 04 Feb 2015 21:36:13 +0000   prometheus                                           BoundPod                                                        failedSync          {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Error syncing pod, skipping: Could not attach disk: Timeout after 10s
Wed, 04 Feb 2015 21:36:33 +0000   prometheus                                           BoundPod                                                        failedSync          {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Error syncing pod, skipping: Could not attach disk: Timeout after 10s
Wed, 04 Feb 2015 21:36:43 +0000   prometheus                                           BoundPod                                                        failedSync          {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Error syncing pod, skipping: Could not attach disk: Timeout after 10s
Wed, 04 Feb 2015 21:36:53 +0000   prometheus                                           BoundPod                                                        failedSync          {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Error syncing pod, skipping: Could not attach disk: Timeout after 10s
Wed, 04 Feb 2015 21:37:53 +0000   prometheus                                           BoundPod            spec.containers{prometheus}                 created             {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Created with docker id 2bd38f6305b6a714dc23d7ef8d1c7bda004db7827811d665ab49aaaa07e7561f
Wed, 04 Feb 2015 21:37:53 +0000   prometheus                                           BoundPod            spec.containers{prometheus}                 pulled              {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Successfully pulled image "gcr.io/eternal_empire_754/core-prometheus-server:latest"
Wed, 04 Feb 2015 21:37:54 +0000   prometheus                                           BoundPod            spec.containers{prometheus}                 started             {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Started with docker id 2bd38f6305b6a714dc23d7ef8d1c7bda004db7827811d665ab49aaaa07e7561f
Wed, 04 Feb 2015 21:38:44 +0000   prometheus                                           BoundPod            spec.containers{node-exporter}              pulled              {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Successfully pulled image "prom/node-exporter:latest"
Wed, 04 Feb 2015 21:38:44 +0000   prometheus                                           BoundPod            spec.containers{node-exporter}              created             {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Created with docker id 162e94ff66bb22136e53293b6827091fd68ca3ae30f19afa622afda901f0a20f
Wed, 04 Feb 2015 21:38:44 +0000   prometheus                                           BoundPod            spec.containers{node-exporter}              started             {kubelet kubernetes-minion-8.c.eternal-empire-754.internal}    Started with docker id 162e94ff66bb22136e53293b6827091fd68ca3ae30f19afa622afda901f0a20f
```

**Comment by smreed** [comment]
I'm going to try with a different container/pod just to be sure.

**Comment by brendandburns** [comment]
There are a couple more PRs in flight that might help fix this:

https://github.com/GoogleCloudPlatform/kubernetes/pull/4093

and

https://github.com/GoogleCloudPlatform/kubernetes/pull/4076

#4093 is merged into head

#4076 is on final approach...

--brendan

**Comment by smreed** [comment]
Those look nice, I'll push a build from master once #4076 lands in order to test this.

**Comment by smreed** [comment]
Nearly forgot, this does not seem dependent on the container I use.

**Comment by smreed** [comment]
Building `kubelet` off of master as I write this, and I'll copy it out to my minions (currently on 0.10.0) in a bit to test this out again. LMK if you'd rather have me test in a brand new cluster.

**Comment by brendandburns** [comment]
sgtm, let me know how it goes.

our e2e has been much cleaner with these two changes submitted...

i'm likely to cut them into a 0.10.1 today, esp. if we get confirmation
from you that it fixes your problems.

thanks
--brendan

**Comment by smreed** [comment]
Success! Just deleted a pod, the disk was properly detached and unmounted. Re-created the pod and it ended up on a different minion, disk was attached correctly with all data intact.

@brendandburns

**Comment by brendandburns** [comment]
Awesome, great to hear (and sorry for the problems)

I will cut this into a 0.10.1 (along with
https://github.com/GoogleCloudPlatform/kubernetes/pull/4169 when it is
merged)

--brendan

**Comment by brendandburns** [comment]
Closing this for now, as user reports it fixed, please re-open if it re-occurs, and I will try to incorporate data read/write into a more extensive e2e test.

**Comment by chromakode** [comment]
I'm seeing this behavior on two clusters I created today on Google Container Enginer. Deleting a pod manually does not detach the disk and deletes the files.

Blew away all of my data -- I was migrating from a persistentDisk in an older cluster. :scream: Luckily I had a backup.

```
$ gcloud preview container kubectl -n test-cluster version
Client Version: version.Info{Major:"0", Minor:"9+", GitVersion:"v0.9.1-dirty", GitCommit:"3623a01bf0e90de6345147eef62894057fe04b29", GitTreeState:"dirty"}
Server Version: &version.Info{Major:"0", Minor:"10+", GitVersion:"v0.10.0-dirty", GitCommit:"71e26cbeb9c0bf1b6cc4c29b1f9d930e53513911", GitTreeState:"dirty"}
```

**Comment by brendandburns** [comment]
Ugh!  Sorry about that!  Is this on 10.1?

I'll take another splunk through the code.  I was fairly certain we now
guard the remove call sites...

Brendan

**Comment by chromakode** [comment]
The server is running 71e26cbeb9c0bf1b6cc4c29b1f9d930e53513911 (which is what Google Container Engine created this evening), which appears to be after #3967 but doesn't include the fixes in #4093 and #4076. This is affecting my new clusters in production on GCE.

**Comment by brendandburns** [comment]
Ok, yeah we are in the process of getting 10.1 rolled out to GKE.  I hope
it will be out today.

Brendan

**Comment by vardaofthevalier** [comment]
I'm currently experiencing this problem as well.  I've tried "kubectl stop ... " and "kubectl delete ..." and had all of my persistent data erased when the pod that mounted it was stopped using either command.  Here is the output of "gcloud alpha container kubectl version" on my local system:

    Client Version: version.Info{Major:"0", Minor:"15", GitVersion:"v0.15.0", GitCommit:"831f3e60d7cd64c61a775d6c78acce1673dd8aa9", GitTreeState:"clean"}
    Server Version: version.Info{Major:"0", Minor:"16+", GitVersion:"v0.16.0-1-gfc76242f4afd2c", GitCommit:"fc76242f4afd2c7e41ae943f4f3f518f3df9377e", GitTreeState:"clean"}

**Comment by swagiaal** [comment]
@vardaofthevalier Hmm.. I was not able to reproduce this. Is it possible you are running into the issue fixed in #7503?
